src/util/strencodings.h

```diff
@@ -369,4 +371,78 @@ std::string Capitalize(std::string str);
  */
 std::optional<uint64_t> ParseByteUnits(std::string_view str, ByteUnit default_multiplier);
 
+namespace util {
+/** consteval version of HexDigit() without the lookup table. */
+consteval uint8_t ConstevalHexDigit(const char c)
+{
+    if (c >= '0' && c <= '9') return c - '0';
+    if (c >= 'a' && c <= 'f') return c - 'a' + 0xa;
+
+    throw "Only lowercase hex digits are allowed, for consistency";
+}
+
+/**
+ * ""_hex is a compile-time user-defined literal returning a
+ * `std::array<std::byte>`, equivalent to ParseHex(). Variants provided:
+ *
+ * - ""_hex_v: Returns `std::vector<std::byte>`, useful for heap allocation or
+ *   variable-length serialization.
+ *
+ * - ""_hex_u8: Returns `std::array<uint8_t>`, for cases where `std::byte` is
+ *   incompatible.
+ *
+ * - ""_hex_v_u8: Returns `std::vector<uint8_t>`, combining heap allocation with
+ *   `uint8_t`.
+ *
+ * @warning It may be necessary to use _v variants when serializing, or vice
+ *   versa, because vectors are assumed to be variable-length and serialized
+ *   with a size prefix, while arrays are considered fixed length and serialized
+ *   with no prefix.
+ *
+ * @warning It may be preferable to use _v variants to save stack space when
+ *   declaring local variables if hex strings are large. Alternately variables
+ *   could be declared constexpr to avoid using stack space.
+ *
+ * @warning Avoid _u8 variants when not necessary, as the codebase migrates to
+ *   use `std::byte` instead of `unsigned char` and `uint8_t`.
+ *
+ * @note The reason ""_hex uses `std::array` is because a heap-based
+ *   `std::vector` cannot cross the compile-time/runtime barrier.
+ */
+inline namespace hex_literals {
+namespace detail {
+
+template <size_t N>
+struct Hex {
```

Comment:
Interesting idea!

Should move it out of the `detail` namespace. That makes the naming of `Hex` more important. Are people still happy with it?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1728787408

This is an interesting idea. I kind of think now that we have support for these suffixes, we should just add an `_rhex` reverse hex suffix and drop this constructor entirely, avoid the footgun of having two constructors where one reverses its input and the other does not.

Also, IMO if adopting this suggestion it would be a little better to keep detail namespace than to expose Hex class as something that would be reasonable to use in other places. But anything should be fine.

Reply:
Not sure about how `_rhex` would work, if it returned `std::array<std::byte, N>` one could accidentally substitute for `_hex`. Keeping `Hex` in `detail` for now though.

Reply:
Think there may still be something to @stickies-v intuition here. Maybe the `consteval` `base_blob`-ctor could somehow be implemented in terms of `""_hex_u8` reverse-copied into `base_blob::m_data` instead of only borrowing `ConstevalHexDigit`. Don't have the stamina to reconcile `Hex<N>` with `std::string_view` right now though.

Reply:
I also like the suggestion, but there are other hex conversion methods in uint256 which could use some compile-time love - I'm fine with doing it in a separate PR.

Reply:
I'm happy to keep things as is for now, this can be considered in follow-ups indeed.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1728787408

The reason I think it would be good to drop this constructor later is that after this PR you can write both:

```c++
uint256{"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"} // bytes are reversed
uint256{"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"_hex_u8} // bytes are not reversed
```

where one constructor reverses the bytes and the other constructor does not reverse the bytes, and the reverse constructor accepts an unlabeled string with no suffix. I think more ideally, we would have just one constructor, instead of two constructors with opposite behaviors.

Supporting an `_rhex` suffix would be one way to have a single constructor and make the argument explicit:

```c++
uint256{"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"_rhex} // bytes are reversed
uint256{"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"_hex} // bytes are not reversed
```

but there could be other ways like adding a `constexpr` Reverse function:

```c++
uint256{Reverse("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"_hex)} // bytes are reversed
uint256{"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"_hex} // bytes are not reversed
```

or supporting string prefixes instead of suffixes:

```c++
uint256{"<ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"_hex)} // bytes are reversed
uint256{">ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"_hex} // bytes are not reversed
```

I don't feel strongly about these ideas but I think they would all simplify the uint256 class and make it more obvious when bytes are being reversed.

Reply:
I like the `_rhex` vs `_hex` suggestion - it's easy to navigate to the definition in the source code and check the difference.
It's a lot more awkward to check it out with an in-string prefix - since it starts with a non-standard, non-hex, arbitrary character.

Reply:
Makes me think maybe `_hex[_v][_u8]` should be changed to return something else than raw `std::array`/`std::vector`. Perhaps some thin wrapper struct that can be implicitly converted to the inner type, but can be caught in the `uint256`-ctors so that `ForwardHex` is explicitly deleted (or handled, reversed), but `ReverseHex` is allowed and treated as if the `uint256(Span<const unsigned char>)`-ctor had been called.